torch/distributed/_tensor/_redistribute.py

```diff
@@ -97,43 +87,135 @@ def _gen_transform_infos(
         else:
             mesh_dims_to_logical_shape.append(current_logical_shape)
 
-        if isinstance(dst, Shard):
-            dst_dim_counts[dst.dim] = dst_dim_counts.get(dst.dim, 0) + 1
-
-        if (
-            isinstance(src, Shard)
-            and isinstance(dst, Shard)
-            and (mesh_ndim > 1 or src_dim_counts[src.dim] != dst_dim_counts[dst.dim])
-        ):
-            # for the case when mesh ndim > 1 or shard dim counts are different
-            # TODO: see if we can optimize the mesh_ndim > 1 case
-            # decompose Shard(i) -> Shard(j) into Shard(i) -> Replicate() -> Shard(j)
-            transform_infos.append(
-                _TransformInfo(
-                    mesh_dim=i,
-                    src_dst_placements=(src, Replicate()),
-                    logical_shape=mesh_dims_to_logical_shape[i],
-                )
-            )
-            transform_infos.append(
-                _TransformInfo(
-                    mesh_dim=i,
-                    src_dst_placements=(Replicate(), dst),
-                    logical_shape=mesh_dims_to_logical_shape[i],
-                )
-            )
-        else:
+    if src_spec.num_shards > 1:
+        # If src_spec have sharding, it could potentially have sharding that is misaligned with dst_spec
+        # a common case of this is nested sharding (i.e. (S(0), S(0)) -> (R, S(0))).
+        # In those cases, we first traverse from inner placement to outer placement
+        # to detect misaligned shardings and properly replicate nested sharding
+        search_transform_infos(
+            src_placements,
+            dst_placements,
+            device_mesh.ndim - 1,
+            device_mesh.ndim,
+            mesh_dims_to_logical_shape,
+            transform_infos,
+            left_to_right=False,
+        )
+
+    # we always traverse from outer placement to inner placement to generate the transform infos
+    search_transform_infos(
+        src_placements,
+        dst_placements,
+        0,
+        device_mesh.ndim,
+        mesh_dims_to_logical_shape,
+        transform_infos,
+    )
+
+    return transform_infos
+
+
+def reshardable_from_src_to_dst(
+    current_placements, target_placements, mesh_dim
+) -> bool:
+    # util function to check if we can redistribute from current placements to target placements
+    # on the current mesh_dim
+    current = current_placements[mesh_dim]
+    target = target_placements[mesh_dim]
+
+    # For current that is Shard, check if it's the inner most sharding
+    # if not then it's not reshardable, else we check the target sharding
+    if current.is_shard():
```

Comment:
In case `[S(1), S(0), S(0)]` => `[R, S(0), S(0)]` would the following happen?

`[S(1), S(0), S(0)]` -> `[S(1), S(0), S(0)]` (`mesh_dim` = 2)
`[S(1), S(0), S(0)]` -> `[S(1), R, S(0)]` (`mesh_dim` = 1) as `reshardable_from_src_to_dst` returns `False` here
`[S(1), R, S(0)]` -> `[R, R, S(0)]` (`mesh_dim` = 0)
Again dim 2 becomes strided sharding.

Ideally there should be only one resharding at dim 0.

Reply:
It seems to me that if we remove this check, the algorithm would be correct, although the method `reshardable_from_src_to_dst` itself won't be as general as you designed it for. In some sense I feel we don't need general check here, since we are only doing right-to-left, and left-to-right, and in left-to-right, we never need to check.

Reply:
OK i think this make sense, let me refactor the current code to make it less general, and just stick with two pass of traversal

Reply:
updated